eventuals/concurrent-ordered.h

```diff
@@ -0,0 +1,254 @@
+#include <map>
+#include <optional>
+#include <tuple>
+
+#include "eventuals/callback.h"
+#include "eventuals/compose.h"
+#include "eventuals/concurrent.h"
+#include "eventuals/flat-map.h"
+#include "eventuals/interrupt.h"
+#include "eventuals/iterate.h"
+#include "eventuals/map.h"
+#include "eventuals/stream.h"
+#include "eventuals/terminal.h"
+
+/////////////////////////////////////////////////////////////////////
+
+namespace eventuals {
+
+/////////////////////////////////////////////////////////////////////
+
+namespace detail {
+
+/////////////////////////////////////////////////////////////////////
+
+struct _ReorderAdaptor {
+  template <typename K_, typename Value_>
+  struct Continuation : public TypeErasedStream {
+    Continuation(K_ k)
+      : k_(std::move(k)) {}
+
+    void Begin(TypeErasedStream& stream) {
+      upstream_ = &stream;
+      k_.Begin(*this);
+    }
+
+    template <typename Value>
+    void Body(std::tuple<int, std::optional<Value>>&& tuple) {
+      CHECK(!done_);
+      int i = std::get<0>(tuple);
+      if (i < 0) {
+        ended_[i * -1] = true;
+        Next();
+      } else if (index_ == i) {
+        CHECK(buffer_[i].empty());
+        k_.Body(std::move(std::get<1>(tuple).value()));
+      } else {
+        CHECK(index_ < i);
+        buffer_[i].push_back(std::move(std::get<1>(tuple).value()));
+        upstream_->Next();
+      }
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Register(Interrupt& interrupt) {
+      k_.Register(interrupt);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    void Ended() {
+      k_.Ended();
+    }
+
+    void Next() override {
+      if (!buffer_[index_].empty()) {
+        auto value = buffer_[index_].front();
+        buffer_[index_].pop_front();
+        k_.Body(std::move(value));
+      } else if (ended_[index_]) {
+        CHECK(buffer_[index_].empty());
+        index_++;
+        Next();
+      } else {
+        upstream_->Next();
+      }
+    }
+
+    void Done() override {
+      done_ = true;
+      buffer_.clear();
+      ended_.clear();
+      upstream_->Done();
+    }
+
+    K_ k_;
+
+    TypeErasedStream* upstream_ = nullptr;
+
+    std::map<int, std::deque<Value_>> buffer_;
+
+    int index_ = 1;
+
+    std::map<int, bool> ended_;
+
+    bool done_ = false;
+  };
+
+  template <typename Value_>
+  struct Composable {
+    template <typename>
```

Comment:
You can use `std::tuple_element` to extract `std::optional<Value>` from the type passed into you, aka, `std::tuple<int, std::optional<Value>>` and then you won't need to pass `Value_` into `ReorderAdaptor`. You can then get `Value` from `std::optional<Value>` via `::value_type`.